File: baseble/src/main/java/com/fcbox/ble/baseble/model/BluetoothLeDeviceStore.java
package com.fcbox.ble.baseble.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 设备信息集合
 */
@SuppressWarnings("unused")
public class BluetoothLeDeviceStore {

    private final Map<String, BluetoothLeDevice> mDeviceMap;

    public BluetoothLeDeviceStore() {
        mDeviceMap = new HashMap<>();
    }

    public void addDevice(BluetoothLeDevice device) {
        if (device == null) {
            return;
        }
        if (mDeviceMap.containsKey(device.getAddress())) {
            mDeviceMap.get(device.getAddress()).updateRssiReading(device.getTimestamp(), device.getRssi());
        } else {
            mDeviceMap.put(device.getAddress(), device);
        }
    }

    public void removeDevice(BluetoothLeDevice device) {
        if (device == null) {
            return;
        }
        mDeviceMap.remove(device.getAddress());
    }

    public void clear() {
        mDeviceMap.clear();
    }

    public Map<String, BluetoothLeDevice> getDeviceMap() {
        return mDeviceMap;
    }

    public List<BluetoothLeDevice> getDeviceList() {
        final List<BluetoothLeDevice> methodResult = new ArrayList<>(mDeviceMap.values());

        Collections.sort(methodResult, new Comparator<BluetoothLeDevice>() {

            @Override
            public int compare(final BluetoothLeDevice arg0, final BluetoothLeDevice arg1) {
                return arg0.getAddress().compareToIgnoreCase(arg1.getAddress());
            }
        });

        return methodResult;
    }

    @Override
    public String toString() {
        return "BluetoothLeDeviceStore{" +
                "DeviceList=" + getDeviceList() +
                '}';
    }
}

File: baseble/src/main/java/com/xzy/ble/baseble/model/BluetoothLeDeviceStore.kt
package com.xzy.ble.baseble.model

import java.util.ArrayList
import java.util.Comparator
import java.util.HashMap
import java.util.Objects

/**
 * 设备信息集合
 */
@Suppress("unused")
class BluetoothLeDeviceStore {

    private val mDeviceMap: MutableMap<String, BluetoothLeDevice>

    val deviceMap: Map<String, BluetoothLeDevice>
        get() = mDeviceMap

    val deviceList: List<BluetoothLeDevice>
        get() {
            val methodResult = ArrayList(mDeviceMap.values)

            methodResult.sortWith(Comparator { arg0, arg1 -> arg0.address.compareTo(arg1.address, ignoreCase = true) })

            return methodResult
        }

    init {
        mDeviceMap = HashMap()
    }

    fun addDevice(device: BluetoothLeDevice?) {
        if (device == null) {
            return
        }
        if (mDeviceMap.containsKey(device.address)) {
            Objects.requireNonNull<BluetoothLeDevice>(mDeviceMap[device.address]).updateRssiReading(device.timestamp, device.rssi)
        } else {
            mDeviceMap[device.address] = device
        }
    }

    fun removeDevice(device: BluetoothLeDevice?) {
        if (device == null) {
            return
        }
        mDeviceMap.remove(device.address)
    }

    fun clear() {
        mDeviceMap.clear()
    }

    override fun toString(): String {
        return "BluetoothLeDeviceStore{" +
                "DeviceList=" + deviceList +
                '}'.toString()
    }
}
